File: src/Algorithms/Lists/Highest_Product_Of_Integers/highest_product_of_three_integers.py

def max_product_of_integers(list_of_ints):
    highest = max(list_of_ints[0], list_of_ints[1])
    lowest = min(list_of_ints[0], list_of_ints[1])

    max_product_of_two = min_product_of_two = highest * lowest

    max_product = max_product_of_two * list_of_ints[2]

    for current in list_of_ints[2:]:
        max_product = max(max_product, max_product_of_two *
                          current, min_product_of_two * current)

        max_product_of_two = max(
            max_product_of_two, highest * current, lowest * current)

        min_product_of_two = min(
            min_product_of_two, highest * current, lowest * current)

        highest = max(highest, current)
        lowest = min(lowest, current)

    return max_product

File: src/Algorithms/Lists/Highest_Product_Of_Integers/highest-product-of-three-integers.ts
// O(n) time
// O(1) space
const solution = (arrayOfInts: number[]) => {
  if (arrayOfInts.length < 3) {
    throw new Error("Less than 3 items!");
  }

  let [lowest, highest] = arrayOfInts.slice(0, 2).sort();

  let lowestProductOfTwo = lowest * highest;
  let highestProductOfTwo = lowestProductOfTwo;
  let maxProduct = highestProductOfTwo * arrayOfInts[2];

  for (let i = 2; i < arrayOfInts.length; i++) {
    const current = arrayOfInts[i];
    maxProduct = Math.max(
      maxProduct,
      lowestProductOfTwo * current,
      highestProductOfTwo * current
    );

    lowestProductOfTwo = Math.min(
      lowestProductOfTwo,
      current * lowest,
      current * highest
    );

    highestProductOfTwo = Math.max(
      highestProductOfTwo,
      current * lowest,
      current * highest
    );

    lowest = Math.min(lowest, current);
    highest = Math.max(highest, current);
  }

  return maxProduct;
};

export const solutions = {
  Solution: solution,
};
